src/rez/solver.py

```diff
@@ -21,16 +23,31 @@
 from rez.vendor.pygraph.algorithms.accessibility import accessibility
 from rez.exceptions import PackageNotFoundError, ResolveError, \
     PackageFamilyNotFoundError, RezSystemError
-from rez.version import VersionRange
+from rez.version import Version, VersionRange
 from rez.version import VersionedObject, Requirement, RequirementList
+from rez.utils.typing import SupportsLessThan, Protocol
 from contextlib import contextmanager
 from enum import Enum
 from itertools import product, chain
+from typing import Any, Callable, Generator, Iterator, TypeVar, TYPE_CHECKING
 import copy
 import time
 import sys
 import os
 
+if TYPE_CHECKING:
+    from rez.resolved_context import ResolvedContext
+    from rez.package_filter import PackageFilterBase
+    from rez.package_order import PackageOrder
+
+
+T = TypeVar("T")
+
+
+class SupportsWrite(Protocol):
```

Comment:
Would something like `io.RawIOBase` be more appropriate here?

Reply:
That could work, but it adds requirements that we don't actually need:  `read`, `readinto`, and `readall` methods.  In our case we only care about writing, so this protocol is more technically correct, since today a user could create any kind of object to write into it as long as it defines a `write` method.